File: my_problems/day39-bit-manipulations-1/hw_q3_DifferentBitsSumPairwise.js
/*****
 * Problem Description

We define f(X, Y) as number of different corresponding bits in binary representation of X and Y.
For example, f(2, 7) = 2, since binary representation of 2 and 7 are 010 and 111, respectively. The first and the third bit differ, so f(2, 7) = 2.

You are given an array of N positive integers, A1, A2 ,..., AN. Find sum of f(Ai, Aj) for all pairs (i, j) such that 1 ≤ i, j ≤ N. Return the answer modulo 109+7.

Problem Constraints
1 <= N <= 10^5
1 <= A[i] <= 2^31 - 1

Input Format
First and only argument of input contains a single integer array A.

Output Format
Return a single integer denoting the sum.

Example Input
Input 1:
 A = [1, 3, 5]
Input 2:
 A = [2, 3]

Example Output
Output 1:
 8
Output 2:
 2

Example Explanation
Explanation 1:
 f(1, 1) + f(1, 3) + f(1, 5) + f(3, 1) + f(3, 3) + f(3, 5) + f(5, 1) + f(5, 3) + f(5, 5) 
 = 0 + 1 + 1 + 1 + 0 + 2 + 1 + 2 + 0 = 8
Explanation 2:
 f(2, 2) + f(2, 3) + f(3, 2) + f(3, 3) = 0 + 1 + 1 + 0 = 2
 */

/**
 * Link: https://www.geeksforgeeks.org/sum-of-bit-differences-among-all-pairs/
 * Efficient Solution – 
 * An Efficient Solution can solve this problem in O(n) time 
 * using the fact that all numbers are represented using 32 bits (or some fixed number of bits). 
 * The idea is to count differences at individual bit positions. 
 * We traverse from 0 to 31 and count numbers with i’th bit set. 
 * Let this count be ‘count’. There would be “n-count” numbers with i’th bit not set. 
 * So count of differences at i’th bit would be “count * (n-count) * 2”, 
 * the reason for this formula is as every pair having one element which has set bit at i’th position 
 * and second element having unset bit at i’th position contributes exactly 1 to sum, 
 * therefore total permutation count will be count*(n-count) 
 * and multiply by 2 is due to one more repetition of all this type of pair as per given condition for making pair 1<=i, j<=N.
 * 
 * TC: O(N)
 * SC: O(1)
 * @param {Array} A array of integers
 * @returns 
 */
const differentBitsSumPairWise = (A) => {

    const N = A.length;
    const M = 1000000007;
    let answer = 0;

    for (let i = 0; i < 32; i++) {

        // count ith set bits for each element in array
        let count = 0;
        for (let j = 0; j < A.length; j++) {
            if (A[j] & (1 << i)) {
                count++;
            }
        }

        // calculate the contribution of ith bit
        const contribution = count * (N - count) * 2;
        // add contribution value to answer
        answer = ((answer % M) + (contribution % M)) % M;

    }
    console.log(answer);
    return answer;

}

/**
 * Test case 1: [1, 3, 5], expected output => 8
 * Test case 2: [2, 3], expected output => 2
 * Test case 3: [2, 6, 5, 4, 8], expected output => 40
 * Test case 4: [9], expected output => 0
 */
differentBitsSumPairWise([1, 3, 5]);
differentBitsSumPairWise([2, 3]);
differentBitsSumPairWise([2, 6, 5, 4, 8]);
differentBitsSumPairWise([9]);

File: my_problems/day39-bit-manipulations-1/hw_q3_Different_Bits_Sum_Pairwise.java
/* Different Bits Sum Pairwise - find contribution of 1s after XORing every pair*/
public class hw_q3_Different_Bits_Sum_Pairwise {
    public int cntBits(int[] A) {

        // Idea: contribution technique
        long answer = 0;
        long m = 1000000007;
        // for every bit position, find the set and unset bits
        // and consider every pair which can contribute the the final answer
        for (int pos = 0; pos < 32; pos++) {
            int setBits = 0;
            for (int i = 0; i < A.length; i++) {
                if (checkBit(A[i], pos)) {
                    setBits++;
                }
            }
            int unsetBits = A.length - setBits;
            // multiply by 2 as we are considering pair (i,j) and (j,i)
            long contribution = (long) setBits * (long) unsetBits * (long) 2;
            answer = (answer % m + contribution % m) % m;
        }
        return (int) answer;

    }

    private boolean checkBit(int n, int pos) {
        if (((n >> pos) & 1) == 1)
            return true;
        return false;
    }

    public static void main(String[] args) {
        hw_q3_Different_Bits_Sum_Pairwise t1 = new hw_q3_Different_Bits_Sum_Pairwise();
        System.out.println(t1.cntBits(new int[] { 1, 3, 5 }));
        System.out.println(t1.cntBits(new int[] { 2, 3 }));
        System.out.println(t1.cntBits(new int[] { 2, 6, 5, 4, 8 }));
        System.out.println(t1.cntBits(new int[] { 9 }));
    }
}

/* Different Bits Sum Pairwise - Approach */

// Brute Force
// f(X, Y) {
// val = X^Y
// 101 => count set in val = 2
// return countsetbits(2);
// }

// Contribution Technique
// 1, 3, 5 => 8

// 0001
// 0011
// 0101
// -----
// 0

// bit pos 0 1
// 0 0 3 => 3*0 = 0
// 1 2 1 => 2
// 2 2 1 => 2
// 3 3 0 => 0
// ==========================
// 4 * 2